BLUEPRINT.md

```diff
@@ -0,0 +1,335 @@
+
+---
+
+#### 🌊 Stream Response to Client
+
+- Server looks up for the bytes that can be cryptographically verified against the requested multihash, in its known storage
+- Server creates a response to send back to the client, which is shaped by the client request:
+  - a simple `multihash` request-response at a time interaction (similar to [Bitswap](https://specs.ipfs.tech/bitswap-protocol/))
+  - more complex responses, such as [commp](https://spec.filecoin.io/systems/filecoin_files/piece/#section-systems.filecoin_files.piece), [blake3](https://github.com/BLAKE3-team/BLAKE3/), multiblock-responses ([CAR](https://ipld.io/specs/transport/car/carv1/)), etc.
+- Stateless, low-memory footprint:
+  - All `HashStreamer` building blocks are pure reads and streaming - ideal for tiny servers or Lambda Functions
+
+![image](./diagrams/hash-stream-3-streamer.svg)
+
+```js
+import { HashStreamer } from '@hash-stream/streamer'
+import { IndexReader } from '@hash-stream/index/reader'
+import { PackReader } from '@hash-stream/pack/reader'
+
+const indexReader = new IndexReader(indexStore)
+const packReader = new PackReader(packStore)
+const hashStreamer = new HashStreamer(indexReader, packReader)
+
+// Get verifiable blobs associated with the target multihash
+for await (const { multihash, bytes } of hashStreamer.stream(targetMultihash)) {
+  // TODO: Encode data stream to send to the client depending on the request
+}
+
+return new Response(dataStream)
+```
+
+---
+
+#### 😎 Done — Trustless Data, Delivered!
+
+Streamed bytes can be hashed by the client on the fly, in order to ensure they match the requested CID.
+
+- Verified content, no server trust required
+- Minimal moving parts
+- Simple transport (HTTP, gRPC, etc.)
+
+![image](./diagrams/hash-stream-simple.svg)
+
+Client libraries like [`@helia/verified-fetch`](https://github.com/ipfs/helia-verified-fetch/tree/main/packages/verified-fetch) can help fetching and verifying responses easily.
+
+```js
+import { createVerifiedFetch } from '@helia/verified-fetch'
+
+const verifiedFetch = await createVerifiedFetch({
+  gateways: ['https://some-hash-stream-server.com'],
+})
+
+const response = await verifiedFetch(`ipfs://${cid}/`)
+const data = await response.blob()
+```
+
+---
+
+### 🛫 Lifecycle of the Response
+
+> How does the streamer response actually get assembled?
+
+#### 🕵️‍♂️ Find Index Records associated with the multihash
+
+- No need to do "content discover" — just lookup for known Index records for the multihash
+
+![image](./diagrams/hash-stream-1-index.svg)
+
+- **What’s an Index?**
+  - A tiny, server-local, map of `multihash` → `IndexRecord`.
+  - One can think of it like a `directory` or `map` for their content inside storage.
+- There are different **fundamental lenses**, which an indexing system MAY be able to index and enable queries for:
+  - A **Blob** is a set of bytes that is individually addressed by a multihash. It can be stored as is in a given store, or within a Pack.
```

Comment:
a `Blob` has size limits is this like a IPFS `Block`?

Reply:
It can be seen as an IPFS Block, if you consider CAR a Pack, Block a Blob, and containing a cryptographic relationship between all the bytes (a Root CID for a dag for example).

In other contexts, it can be anything, just some bytes that are addressed by a multihash. So there are no limits at this level, only if the underlying storage has some kind of limit per entry